### PR TITLE
agent: Refactor memory management in Profiler class

### PR DESCRIPTION
**Pull request by ikavalio**
Hi,

This PR is big, but it doesn't change any existing functionality. It aims to simplify memory management in `Profiler` (no need to delete things explicitly) and enable profiler configuration test in Travis build (currently it creates a JVM, though technically tests don't need it and ASAN doesn't like it).

Summary of important changes:
1. Replace all `char*` in `ConfigurationOptions` with `std::string`. This allows not to use `safe_copy_string` and `safe_free_string` functions. Though strings are copied rather than just assigned as `char*` it doesn't affect profiler's runtime (time when profiler is actually running), but makes code much simpler.
2. `Processor` owns `CircularQueue` and `SignalHandler`. Processor will manage both buffer and signal handler. Previously signal handling logic was distributed between `Profiler` and `Processor`, but it can be done just by `Processor`. Moreover lifetimes of both buffer and handler are now limited by lifetime of processor.
3. Use `std::unique_ptr` to manage references to `LogWriter` and `Processor` instead of using pointers directly. There's no need to write if's/delete's and explicit destructor.
4. `LogWriter` now owns both `ostream` and `ofstream`. It makes things much easier and allows implementation of e.g. log rotation just within `LogWriter`. Old interface is still available for testing purposes. I also moved `Profiler::lookupFrameInformation` to `LogWriter` since there's no need for it to access `Profiler` specific things.
5. Split `Profiler::handle` logic between `Profiler::handle` and `Processor::handle`. `Profiler` has access to JVM and gets a JNI env, but `Processor` does actual AGCT and pushes info to buffer. This allows to use `Profiler` in tests w/o actual JVM.

All changes are GCC 4.4 compatible.

### CODE REVIEW

File: src/main/cpp/globals.h [hunk 1]
```diff
@@ -33,32 +34,40 @@ const int MAX_FRAMES_TO_CAPTURE = 2048;
...
+            samplingIntervalMax(config.samplingIntervalMax),
+            logFilePath(config.logFilePath),
+            host(config.host),
+            port(config.port),
+            start(config.start),
+            maxFramesToCapture(config.maxFramesToCapture) {
+    }
+
     virtual ~ConfigurationOptions() {
-      if (logFilePath) safe_free_string(logFilePath);
```
Comment (RichardWarburton) [inline src/main/cpp/globals.h]: shouldn't we still delete the std::string objects when the `ConfigurationOptions` is destroyed? Looks like in this PR the std::string objects are effectively owned by `ConfigurationOptions`.

File: src/main/cpp/signal_handler.h [hunk 2]
```diff
@@ -34,14 +35,12 @@ class SignalHandler {
...
-    ~SignalHandler() {
-        delete[] timingIntervals;
-    }
+    ~SignalHandler() {}
 
 private:
     int intervalIndex;
-    int *timingIntervals;
     int currentInterval;
+    std::array<int, NUMBER_OF_INTERVALS> timingIntervals;
```
Comment (RichardWarburton) [inline src/main/cpp/signal_handler.h]: Should we not be deleting the `timingIntervals` std::array when the SignalHandler gets destroyed - looks like the SignalHandler owns the array.

**Comment by RichardWarburton** [comment]
Thanks again for the PR @ikavalio - great stuff. I've just put a couple of questions around the cleaning up of objects before merging.

**Comment by ikavalio** [comment]
Hi @RichardWarburton, thanks for your comments. In both cases everything should be handled correctly and I don't see any errors reported by address/leak sanitizer. 

All `std::string` fields of `ConfigurationOptions` will copy content (in some cases move) from source strings no matter if they are raw `char*` (in agent.cpp copying `std::string::assign` is used) or other `std::string`, so destructor/assignment operator will free occupied memory correctly and transparently. That's why I'm using `std::string`, so we don't need to keep track of the owner of `char*`.

`std::array` in this case is identical to `int timingIntervals[NUMBER_OF_INTERVALS];` since NUMBER_OF_INTERVALS is a compile time constant and it will be freed automatically when new Processor is created (e.g. if user changes sampling interval settings) and current goes out of the scope.

**Comment by RichardWarburton** [comment]
Gotcha, thanks.

**Pull request merged**